libmamba/include/mamba/download/downloader.hpp

```diff
@@ -98,48 +30,49 @@ namespace mamba
         termination_function on_unexpected_termination = std::nullopt;
     };
 
-    class DownloadMonitor
+    class Monitor
     {
     public:
 
-        virtual ~DownloadMonitor() = default;
+        virtual ~Monitor() = default;
 
-        DownloadMonitor(const DownloadMonitor&) = delete;
-        DownloadMonitor& operator=(const DownloadMonitor&) = delete;
-        DownloadMonitor(DownloadMonitor&&) = delete;
-        DownloadMonitor& operator=(DownloadMonitor&&) = delete;
+        Monitor(const Monitor&) = delete;
+        Monitor& operator=(const Monitor&) = delete;
+        Monitor(Monitor&&) = delete;
+        Monitor& operator=(Monitor&&) = delete;
 
-        void observe(MultiDownloadRequest& requests, DownloadOptions& options);
+        void observe(MultiRequest& requests, Options& options);
         void on_done();
         void on_unexpected_termination();
 
     protected:
 
-        DownloadMonitor() = default;
+        Monitor() = default;
 
     private:
 
-        virtual void observe_impl(MultiDownloadRequest& requests, DownloadOptions& options) = 0;
+        virtual void observe_impl(MultiRequest& requests, Options& options) = 0;
         virtual void on_done_impl() = 0;
         virtual void on_unexpected_termination_impl() = 0;
     };
 
-    MultiDownloadResult download(
-        MultiDownloadRequest requests,
+    MultiResult download(
+        MultiRequest requests,
+        const mirror_map& mirrors,
         const Context& context,
-        DownloadOptions options = {},
-        DownloadMonitor* monitor = nullptr
+        Options options = {},
+        Monitor* monitor = nullptr
     );
 
-    DownloadResult download(
-        DownloadRequest request,
+    Result download(
+        Request request,
+        const mirror_map& mirrors,
         const Context& context,
-        DownloadOptions options = {},
-        DownloadMonitor* monitor = nullptr
+        Options options = {},
+        Monitor* monitor = nullptr
```

Comment:
Is optional reference meant here?

Reply:
Yes, but Monitor is a polymorphic type, and we cannot use `std::optional` with `references`.

Reply:
Ah I just posted a review with a similar remark.
`Monitor` being polymorphic is not a problem as long as a reference or pointer is passed. However indeed `std::optional` does not support references (it's being discussed to enable it).
Question: arent we already using `tl::optional` which does support references?

Reply:
Nope, we only use `tl::expected`, and I'm a bit reluctant to add a new dependency just for handling optional references in a few cases :s

Reply:
Ah yes I mixed these up. Ok then.